src/main/java/io/airlift/units/ThreadsCount.java

```diff
@@ -0,0 +1,195 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.units;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.function.Supplier;
+
+import static io.airlift.units.Preconditions.checkArgument;
+import static io.airlift.units.Preconditions.checkState;
+import static java.lang.Math.min;
+import static java.lang.Math.round;
+import static java.lang.Math.toIntExact;
+import static java.lang.String.format;
+import static java.nio.file.Files.exists;
+import static java.nio.file.Files.readAllLines;
+
+public class ThreadsCount
+        implements Comparable<ThreadsCount>
+{
+    private static final String PER_CORE_SUFFIX = "C";
+    private static final Supplier<Integer> AVAILABLE_PROCESSORS = MachineInfo::getAvailablePhysicalProcessorCount;
+    private final int threadsCount;
+
+    ThreadsCount(int threadsCount)
+    {
+        checkArgument(threadsCount >= 0, "Threads count cannot be smaller than 0");
+        this.threadsCount = threadsCount;
+    }
+
+    public int getThreadsCount()
+    {
+        return threadsCount;
+    }
+
+    public ThreadsCount nextPowerOf2()
+    {
+        return new ThreadsCount(nextPowerOfTwo(getThreadsCount()));
+    }
+
+    public static ThreadsCount exactValueOf(int value)
+    {
+        return new ThreadsCount(value);
+    }
+
+    public static ThreadsCount valueOf(String value)
+    {
+        if (value.endsWith(PER_CORE_SUFFIX)) {
+            float parsedMultiplier = parseFloat(value.substring(0, value.lastIndexOf(PER_CORE_SUFFIX)).trim());
+            return new ThreadsCount(round(parsedMultiplier * AVAILABLE_PROCESSORS.get()));
+        }
+
+        return new ThreadsCount(parseInteger(value));
+    }
+
+    public static ThreadsCount boundedValueOf(String value, String minValue, String maxValue)
+    {
+        ThreadsCount parsed = ThreadsCount.valueOf(value);
+        ThreadsCount min = ThreadsCount.valueOf(minValue);
+        ThreadsCount max = ThreadsCount.valueOf(maxValue);
+
+        if (parsed.compareTo(min) < 0) {
+            return min;
+        }
+
+        if (parsed.compareTo(max) > 0) {
+            return max;
+        }
+        return parsed;
+    }
+
+    private static float parseFloat(String value)
+    {
+        try {
+            return Float.parseFloat(value);
+        }
+        catch (NumberFormatException e) {
+            throw new IllegalArgumentException(format("Cannot parse value '%s' as float", value), e);
+        }
+    }
+
+    private static int parseInteger(String value)
+    {
+        try {
+            return Integer.parseInt(value);
+        }
+        catch (NumberFormatException e) {
+            throw new IllegalArgumentException(format("Cannot parse value '%s' as integer", value), e);
+        }
+    }
+
+    static int nextPowerOfTwo(int x)
+    {
+        if (x == 0) {
```

Comment:
We can base this on Guava's `ceilingPowerOfTwo`, which handles zero automatically:
```java
if (x > (1 << 30)) {
    throw new IllegalStateException("Thread count is too large for next power of two: " + x);
}
return 1 << -Integer.numberOfLeadingZeros(x - 1);